src/PgRvn.Server/MessageReader.cs

```diff
@@ -0,0 +1,390 @@
+            msgLen -= sizeof(int);
+
+            var processId = await ReadInt32Async(reader, token);
+            msgLen -= sizeof(int);
+
+            var sessionId = await ReadInt32Async(reader, token);
+            msgLen -= sizeof(int);
+
+            if (msgLen != 0)
+            {
+                throw new PgFatalException(PgErrorCodes.ProtocolViolation,
+                    $"Message is bigger than specified in msgLen field - {msgLen} extra bytes in message.");
+            }
+
+            return new Cancel
+            {
+                ProcessId = processId,
+                SessionId = sessionId
+            };
+        }
+
+        public async Task<Message> ReadMessage(PipeReader reader, CancellationToken token)
+        {
+            var msgType = await ReadByteAsync(reader, token);
+            var msgLen = await ReadInt32Async(reader, token) - sizeof(int);
+
+            Message message;
+            switch (msgType)
+            {
+                case (byte)MessageType.Parse:
+                    (message, msgLen) = await ReadParse(msgLen, reader, token);
+                    break;
+
+                case (byte)MessageType.Bind:
+                    (message, msgLen) = await ReadBind(msgLen, reader, token);
+                    break;
+
+                case (byte)MessageType.Describe:
+                    (message, msgLen) = await ReadDescribe(msgLen, reader, token);
+                    break;
+
+                case (byte)MessageType.Execute:
+                    (message, msgLen) = await ReadExecute(msgLen, reader, token);
+                    break;
+
+                case (byte)MessageType.Sync:
+                    message = new Sync();
+                    break;
+
+                case (byte)MessageType.Terminate:
+                    message = new Terminate();
+                    break;
+
+                case (byte)MessageType.Query:
+                    (message, msgLen) = await ReadQuery(msgLen, reader, token);
+                    break;
+                case (byte)MessageType.Close:
+                    (message, msgLen) = await ReadClose(msgLen, reader, token);
+                    break;
+                case (byte)MessageType.Flush:
+                    message = new Flush();
+                    break;
+                default:
+                    // TODO: Catch cases of messages that are real but *unsupported* and skip them if we can instead of fatally exiting
+                    throw new PgFatalException(PgErrorCodes.ProtocolViolation, "Message type unrecognized: " + (char)msgType);
+            }
+
+            if (msgLen != 0)
+            {
+                throw new PgFatalException(PgErrorCodes.ProtocolViolation,
+                    $"Message is larger than specified in msgLen field, {msgLen} extra bytes in message.");
+            }
+
+            return message;
+        }
+
+        private async Task<(Parse, int)> ReadParse(int msgLen, PipeReader reader, CancellationToken token)
```

Comment:
I don't think it would look nice, `Message.Length` will never be used outside this scope, so I think it's fine that we return it this way. The `msgLen` that we return is basically bytesRead.

I modified the code a bit to reflect that.

Reply:
okay